### ISSUE TITLE
it support opencl1.1 device？

### ISSUE THREAD

**Issue by zif520** [amd/OpenCL-caffe#17]
i want to use zhe opencl caffe in android ,but only support opencl 1.1，can i use it but only chang little function？

**Comment by naibaf7** [comment]
@zif520 
It could potentially be made to work with android, but I imagine that will be quite some work, starting with getting it to compile as a library first.

You can try, but I don't have this planned as a feature in the near future, so I'm not coding for this feature.

**Comment by gujunli** [comment]
hmm this is an interesting question. we did not try in OpenCL1.1, the first step will be figuring out what has been new OpenCL1.2 vs 1.1, and removed the new functions of 1.2

**Comment by hughperkins** [comment]
in case it's useful, I know that clBLAS 2.4 worked ok in 1.1, back in the days when NVIDIA only supported 1.1

1.1 is quite old now, and recent Adrenos support not just 1.2 but 2.0 https://en.wikipedia.org/wiki/Adreno

**Comment by zif520** [comment]
thanks for all ,
i found most of mali GPU only supply opencl 1.1 ,i also refer to https://github.com/naibaf7/caffe to do it ,perhaps i will complete it some days ,but there is some problems now.

**Comment by hughperkins** [comment]
Sounds cool.  Just out of curiosity, which 1.2 functions are being used?

**Comment by zif520** [comment]
@hughperkins 
we only use opencl 1.1 ,and now it is worked ok with ViennaCL ,but there are a little slow,so i want to try clblas now. my telephone is mate8 mali T880

**Comment by hughperkins** [comment]
Sounds good. You have two approaches really:
- insert shims in between clblas et al and opencl (this is what I would do)
- modify clblas et al (this sounds... harder, but hard to say without checking first)

To insert shims, what I would do is:
- download/build clew https://github.com/hughperkins/clew   This is like a thin layer in between clblas etc, and opencl.  perfect for adding shims
- give clew library to clblas etc.  Tell clblas it is the opencl library. It will believe you :-)
- stub out appropriate opencl 1.2 function calls in clew, so that they are implemented using 1.1 functions and / or custom opencl kernels

I would think you wouldnt want to directly modify the clew library, but create a new library (call it opencl-1.2-on-1.1 for example), which will stub out the clew methods, ie by overwriting the function pointers eg at https://github.com/hughperkins/clew/blob/master/src/clew.c#L112 )

**Comment by zif520** [comment]
@hughperkins 
thanks for your illustration.: ) , i will try it !

**Comment by psyhtest** [comment]
I was able to compile OpenCL-caffe with clBLAS 2.4 (and openBLAS 0.2.17). But what really stopped me in my tracks was this:

```
// src/caffe/device.cpp
 namespace caffe {
 #ifndef CPU_ONLY
 string buildOption = "-x clc++ ";
```

and then this:

```
// src/caffe/ocl/util.cl
#pragma OPENCL EXTENSION cl_amd_printf : enable

template <class T>
__kernel void OCL_memset(__global T* buffer, const T value, const int size, const int buf_offset) {
```

C'mon, using templates with OpenCL 1.x?

**Comment by hughperkins** [comment]
:-D

**Comment by naibaf7** [comment]
@psyhtest 
Here this one has OpenCL 1.1 support: https://github.com/BVLC/caffe/tree/opencl

**Comment by psyhtest** [comment]
@naibaf7 Yes, I'm using your port now - and it's much more promising! :+1: I hope to provide you some contributions in due course.

@zif520 I used to lead the OpenCL compiler team for the ARM Mali GPUs, so I'm always happy to see people using Mali. :+1: I do not, however, think you can get good performance with clBLAS. I plan to connect another BLAS library that can be tuned for embedded GPUs.

**Comment by hughperkins** [comment]
> I plan to connect another BLAS library that can be tuned for embedded GPUs.

which library is this?

**Comment by psyhtest** [comment]
@hughperkins https://github.com/CNugteren/CLBlast

**Comment by hughperkins** [comment]
Thanks!

**Comment by soulslicer** [comment]
So what version of OpenCL does this actually support? Does it support 1.2?
